nvmeof: fix bug "already connected"

gadididi opened the pull request:
# Describe what this PR does #

This PR fixes the NVMe-oF CSI node plugin's handling of already-connected subsystems. Previously, when `NodeStageVolume` was called for a volume, whose subsystem was already connected, the `nvme connect` command would return "already connected" error, causing the staging operation to fail.

The fix introduces a pre-connection check that calls `nvme list-subsys -o json` once before attempting connections. For each gateway listener, the driver checks if a live path already exists from the specific HostNQN to that gateway IP:port. If found, the connection attempt is skipped, and the driver proceeds directly to device path resolution and staging.

## Is there anything that requires special attention ##

**Backward Compatibility**: This change is fully backward compatible. The behavior only changes in the error case (already connected), converting it to a success case.


**Key behavioral changes**:
- The driver now calls `nvme list-subsys -o json` once per `ConnectSubsystem` call to retrieve existing connections
- Connection attempts are skipped (and marks `success = true` for the `ConnectSubsystem()` ) only when a live path exists to the specific gateway IP:port



## Related issues ##
example of the bug:

```bash 
E1126 12:07:47.541273  179952 cephcmds.go:156] ID: 58 Req-ID: 0001-0011-openshift-storage-0000000000000002-c7a2646a-c39e-44b7-8c2a-287c8c79c1c5 an error (exit status 1) and stderror (already connected
) occurred while running nvme args: [connect -t tcp -n nqn.2016-06.io.ceph:subsystem.test-integration -a d.c.b.a -s 4420 -l 1800 --hostnqn nqn.2014-08.org.nvmexpress:gdidi-9dqf2-worker-1-5f48x]
W1126 12:07:47.541316  179952 nvmeof_initiator.go:113] ID: 58 Req-ID: 0001-0011-openshift-storage-0000000000000002-c7a2646a-c39e-44b7-8c2a-287c8c79c1c5 Failed to connect to d.c.b.a:4420 - stdout: , stderr: already connected
I1126 12:07:47.541338  179952 nvmeof_initiator.go:92] ID: 58 Req-ID: 0001-0011-openshift-storage-0000000000000002-c7a2646a-c39e-44b7-8c2a-287c8c79c1c5 Connecting to NVMe-oF subsystem nqn.2016-06.io.ceph:subsystem.test-integration at a.b.c.d:4420
E1126 12:07:47.547762  179952 cephcmds.go:156] ID: 58 Req-ID: 0001-0011-openshift-storage-0000000000000002-c7a2646a-c39e-44b7-8c2a-287c8c79c1c5 an error (exit status 1) and stderror (already connected
) occurred while running nvme args: [connect -t tcp -n nqn.2016-06.io.ceph:subsystem.test-integration -a a.b.c.d -s 4420 -l 1800 --hostnqn nqn.2014-08.org.nvmexpress:gdidi-9dqf2-worker-1-5f48x]
W1126 12:07:47.547797  179952 nvmeof_initiator.go:113] ID: 58 Req-ID: 0001-0011-openshift-storage-0000000000000002-c7a2646a-c39e-44b7-8c2a-287c8c79c1c5 Failed to connect to a.b.c.d:4420 - stdout: , stderr: already connected
E1126 12:07:47.547939  179952 utils.go:355] ID: 58 Req-ID: 0001-0011-openshift-storage-0000000000000002-c7a2646a-c39e-44b7-8c2a-287c8c79c1c5 GRPC error: rpc error: code = Internal desc = failed to connect to subsystem: failed to connect to any gateway address for subsystem nqn.2016-06.io.ceph:subsystem.test-integration

```


**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer
  guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release
  notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md)
  updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

gadididi committed:
nvmeof: add helper functions listSubsystems() isPathConnected()

we would like to check and know if specific subsystem is
already connected from specific listener to specifc host

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

gadididi committed:
nvmeof: fix already connected bug

when ConnectSubsystem() is called to connect to an already
connected subsystem listener <--> host, the ExecCommandWithTimeout returns
err "already connected". Instead of parsing the error message,
it is cleaner and more convenient to check the
connection state using isSubsystemConnected().
We always call ConnectSubsystem() with HostNQN,
so calling isPathConnected() without hostnqn does not make sense.
Therefore, the check is placed under if req.HostNQN.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

nixpanic commented on internal/nvmeof/nvmeof_initiator.go:
```diff
@@ -59,6 +61,18 @@ type ConnectRequest struct {
+	Subsystems []struct {
+		NQN   string `json:"NQN"`
+		Paths []struct {
+			Address string `json:"Address"` // Format: "traddr=10.242.64.32,trsvcid=4420,src_addr=10.242.64.33"
```
If you make this a type `nvmeAddress` instead of `string`, you can define a function like
```
func (na *nvmeAddress) UnmarshalJSON(data []byte) error
```
Inside that function you can parse the data, split `traddr` and so on.

gadididi replied:
fixed

nixpanic commented on internal/nvmeof/nvmeof_initiator.go:
```diff
@@ -100,12 +100,35 @@ func (ni *nvmeInitiator) LoadKernelModules(ctx context.Context) error {
 // ConnectSubsystem connects to an NVMe-oF subsystem.
 func (ni *nvmeInitiator) ConnectSubsystem(ctx context.Context, req *ConnectRequest) (bool, error) {
+	// Get existing subsystem connections once to avoid repeated nvme list-subsys calls
+	var existingConnections []nvmeHost
```
you could also create `nvmeConnections` as a type `[]nvmeHost`, that way you can add a function
```go
func (nc *nvmeConnections) hasLivePathToGateway(..)
```

gadididi replied:
fixed

nixpanic reviewed (commented):
looks good, just some suggestions to consider to improve the maintainability in the future.

nixpanic commented on internal/nvmeof/nvmeof_initiator.go:
```diff
@@ -59,6 +61,54 @@ type ConnectRequest struct {
+type nvmeHostConnections []nvmeHost
+
+// UnmarshalJSON implements custom JSON unmarshaling for nvmePathAddress.
+func (na *nvmePathAddress) UnmarshalJSON(data []byte) error {
```
nice!

gadididi replied:
thanks 🙂

nixpanic commented on internal/nvmeof/nvmeof_initiator.go:
```diff
@@ -152,3 +202,45 @@ func (ni *nvmeInitiator) GetNamespaceDeviceByUUID(ctx context.Context, uuid stri
+	}
+
+	var hosts nvmeHostConnections
+	if err := json.Unmarshal([]byte(stdout), &hosts); err != nil {
```
The parsing like this allows for a nice unit-test with example json input. Would be good to have that as well. Can be done in a separate PR, but ideally is included in this one.

gadididi replied:
added

nixpanic commented on internal/nvmeof/nvmeof_initiator.go:
```diff
@@ -38,6 +38,8 @@ const (
 	disconnectTimeout = 15 * time.Second
 )
 
+const nvmeConnectTimeout = "1800" // Connection timeout in seconds
```
is this timeout based on anything, it looks rather long, but maybe there is a good reason for that?

gadididi replied:
Hi, the naming is not accurate. I will change it. 
but the value, 1800 is best practice for nvmeof users. 
This param defines how long (in seconds) the kernel will retry reconnecting to a failed controller before giving up.
This is for case of losing network.

gadididi committed:
nvmeof: add unit tests for nvmeHost unmarshel

nvmeHost struct comes from nvmeof list-subsys command output.
we parse the output and seach for already live path.
this commit includes unit tests for some output cases.

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

Madhu-1 commented on internal/nvmeof/nvmeof_initiator.go:
```diff
@@ -152,3 +231,45 @@ func (ni *nvmeInitiator) GetNamespaceDeviceByUUID(ctx context.Context, uuid stri
+
+// listSubsystems retrieves current NVMe subsystem connections.
+func listSubsystems(ctx context.Context) (nvmeHostConnections, error) {
+	stdout, _, err := util.ExecCommandWithTimeout(ctx, 5*time.Second, "nvme", "list-subsys", "-o", "json")
```
is 5 seconds enough when we have more sub system or having some network problem etc, lets keep this at 1 min for now?

gadididi replied:
fixed

Madhu-1 reviewed (dismissed):
LGTM

Madhu-1 commented:
https://github.com/ceph/ceph-csi/pull/5808/commits/5c5f4ffa7f3b32de1011a22fad1bdd5e20fb0487 typo in the commit message :)

nixpanic commented:
Thanks for the additional description/comments and the unit-tests!

gadididi commented:
@Mergifyio queue

gadididi commented:
@mergifyio requeue

gadididi commented:
@Mergifyio queue

gadididi commented:
@mergifyio requeue

nixpanic commented:
We don't have e2e testing for nvmeof yet, no need for e2e at the moment.

nixpanic commented:
@mergifyio queue

gadididi commented:
@mergifyio requeue

gadididi commented:
@mergifyio rebase

gadididi commented:
@Mergifyio queue
